feat: spine 4.1 plugin support

epreston opened the pull request:
Add Spine 4.1 Plugin Support

- Add media and examples for Spine 4.1 runtime
- Update build system to create 4.1 plugin
- Add prebuilt plugins
- Update docs, minor tool adjustments
- Switch to semver version comparison to support 4.1 file versions

Closes #57 
Fixes #52 
Fixes #70 (for 4.1 support, engine version 1.27 and later)

Important changes separated out into commit: 539ff9c85e68e288f58adfd66cab539068183eeb

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

epreston committed:
fix serve command

fix serve command

epreston committed:
Add spine-core": "4.1.36" package files

Use correct package
    "@esotericsoftware/spine-core": "4.1.36",

epreston committed:
Update Build to support 4.1

Name files based on max supported version

epreston committed:
Update spine41.js

Working demo

epreston committed:
Add 4.1 demo

epreston committed:
4.1 plugin builds

draft

epreston committed:
Update README.md

update readme

mvaligursky commented:
Fantastic, thanks for doing this! I'm on holidays, but will check it out later.
We could also update the spine example in the engine examples to use the latest spine version.

mvaligursky commented on README.md:
```diff
@@ -12,13 +12,19 @@ Examples such as the Hero above can be found in the `examples` folder. To run th
+
+| Spine Editor | PlayCanvas Engine |
+| ------------ | ----------------- |
+| 3.6          | Up to 1.65        |
```
Can 3.6 and 3.8 still be used with newer engine?

epreston replied:
Yes, they work with warnings on Mesh see #70.  The rewrite might address this.

epreston replied:
Engine support is a best guess from looking at commit history.

epreston commented:
Happy holidays. I hope to do a rewrite so this builds proper modules next, then the PR for engine examples.  That should resolve the issues you pointed out.

epreston commented:
Demo PlayCanvas Editor project for evaluation / testing following instructions in readme.
https://playcanvas.com/project/1124022/

willeastcott commented:
This is cool. But it's a shame you have to duplicate `spline.js` like that. Can you not just leverage the `versionCompare` function to determine if it's the 4.1 runtime and then do what's required based on that?

epreston commented:
Possibly, there's a few considerations.  

Next step is the rewrite and retool.  I wanted to ensure the identical build products were created for legacy versions.  There's outstanding issues on the old stuff, there's opportunities on the new stuff.  Duplicating that one file allows most of the code to remain shared while de-risking future work and cutting test burden as long as shared files aren't changed.

willeastcott commented:
Yeah, but if I want to refactor/optimize/etc any code in `spline.js`, I'll have to do it in two places now. And there'll no doubt be more versions of Spline we'll have to support in the future. Since there's only 2 or 3 places where the code would diverge for 4.1, I think it's preferable to keep it all in a single file.

willeastcott commented on src/spine41.js:
```diff
@@ -0,0 +1,669 @@
+        this.skeletonVersion = _skeletonData.version;
+        this._spine_3_6_0 = versionCompare(this.skeletonVersion, "3.6.0") <= 0;
+        this._spine_3_7_99 = versionCompare(this.skeletonVersion, "3.7.99") <= 0;
+        this._spine_3_7_99 = versionCompare(this.skeletonVersion, "4.1.23") <= 0;
```
This looks wrong. 😄

epreston replied:
Possibly, the intent is to be using the stuff from that flag "_spine_3_7_9" with first version of the 4.1 file format that became stable.  I think it goes up to "4.1.36".

I'm going for clarity on the exact changes required. The example source they provide for the engines I know well supports much more / checks other things, etc.

willeastcott replied:
Sorry, I'm not quite sure I follow. Line 158 simply overwrites what was written on line 157.

epreston replied:
Your right.  It's because of a deeper bug with the `versionCompare` function.  I'll update.

Short story, it does not support versions like "4.1.23-beta".

Issue:

Spine files can have non numeric version values in compliance with `semver`.  The json file in the 4.1 example has "4.1.23-beta".  `versionCompare` will return `nan` when it's not getting numbers.  The lexical compare option in `versionCompare` is not fully compliant with `semver` so that doesn't sort the issue.

Resolution: 

- switch to `semver` library parsing like the Spine lib.

Details:

The current `versionCompare` is a copy / paste form this stackoverflow post.

https://stackoverflow.com/questions/6832596/how-can-i-compare-software-version-number-using-javascript-only-numbers

But it has a few issues as noted in this gist:

https://gist.github.com/TheDistantSea/8021359

Reference:

https://www.npmjs.com/package/semver#prerelease-identifiers

epreston replied:
Resolved in: [0c72c83](https://github.com/playcanvas/playcanvas-spine/pull/73/commits/0c72c83ea0279190593704a389a4d9aef37cc6a6)

epreston commented:
There should have been a separate file for each file format from the beginning.  That's the least of the structural problems.

Part of complexity and why it took so long to address is because its been done in patches.  Bit early to determine the final structure and maintenance requirements.  I recommend breaking this into pieces and not breaking existing projects / customers that depend on it.

At the moment, you have a workable product using the process that has been done 2 times before.  Happy to build out 4.1 support, add a transpile step to ensure it matches a language standard, then use that to resolve issues with its usage in Engine and Editor.

I'd like to be further along before owning any old issues.

willeastcott commented on src/spine41.js:
```diff
@@ -0,0 +1,669 @@
+
+    Spine.prototype.SubmitBatch = function (indexBase, indexCount, materialKey) {
+        if (indexCount > 0) {
+            var mesh = new pc.Mesh(this._app.graphicsDevice);
```
You seem to be passing the graphics device to the `Mesh` constructor now, but the other copy of spine.js doesn't have this fix. Might even be worth having this in a separate PR since it's unrelated.

epreston replied:
That would break anyone using the old file versions on engine 1.57 and below ?  These file formats go back 7 years, 1.57 was 10 months ago ?  I don't know how far back support is required.

epreston replied:
You want this on 4.1 so it builds and runs without warnings on 1.57 and above.

willeastcott replied:
The `Mesh` constructor switched to accepting a `GraphicsDevice` as a parameter over 3 years ago:

https://github.com/playcanvas/engine/pull/1960/files#diff-f68894866f45f9fc28805640c3670b3865dc50dc44037b6d6fb602fe86726924R144

All `Mesh` constructors across the source should be updated based on the current `Mesh` constructor signature. I don't see how anything could break.

epreston replied:
Your right !  The last commit was about line endings not changing the signature.  Yeah, lets bring it in when we refactor.  #70

epreston commented:
What's the direction on this one ?  What it all in one file ?

What's the specification ?

epreston committed:
Add support for building minimal semver iife

Existing build system combines a number of iife files.  This will create a minimum semver variable in the global namespace.

epreston committed:
Update 4.1 build to include semver

add prebuild "minimal" semver iife
update grunt file to combine it in the correct order

epreston committed:
Update spine41.js

Ensure version comparison supports semver standard as used in the test / validation file provided by Spine.

https://github.com/playcanvas/playcanvas-spine/pull/73#discussion_r1287153169

epreston committed:
Update 4.1 prebuilt modules

epreston commented:
As noted by @willeastcott here: https://github.com/playcanvas/playcanvas-spine/pull/73#discussion_r1285542161

Added correct semver compliant version comparison to support the version attributes found in 4.1 files.

epreston committed:
Update README.md

update supported version information
add note about scripts loading order

willeastcott commented on README.md:
```diff
@@ -12,13 +12,21 @@ Examples such as the Hero above can be found in the `examples` folder. To run th
 
 Create an entity with a script component and add the script `spine` to it. Upload your exported spine resources (atlas, skeleton json file, textures) and attach them to the spine script on your entity.
 
+Ensure the plugin file is listed before the Playcanvas script in the [Scripts Loading Order](https://developer.playcanvas.com/en/user-manual/scripting/loading-order/).
```
```suggestion
Ensure the plugin file is listed before the PlayCanvas script in the [Scripts Loading Order](https://developer.playcanvas.com/en/user-manual/scripting/loading-order/).
```

epreston committed:
Update README.md

fix typo

Co-authored-by: Will Eastcott <will@playcanvas.com>

willeastcott commented:
Great to see this progressing. Just want to highlight some things:

* I'm still keen to merge `spine.js` back into a single file. The code there needs heavy refactoring (update to ES6 class in an ES Module). We don't want to have to do this twice.
* I see you've added an esbuild dependency. At some point, we need to have this repo conform with all other PlayCanvas libs and use Rollup for consistency (replacing Gulp too in the process). I guess when we are on ESM/Rollup, we can just import semver into the code from the NPM module...
* Minor thing but we should switch from `http-server` to `serve` too which we seem to have settled on elsewhere. That can be a separate PR.

epreston commented:
Sounds good.

After the refactor, build replacement, introducing code transformation, you might not even have a `spine.js` file.  The final result should be dry but also not have the developer managing the concerns of all the engine versions and file versions with a bunch of switch statements.

Yes, Rollup is paired with Babel or esbuild to do the transformations and automatic polyfils. The greater issue that sparked this work is to address dependencies like this that are being bundled without transformation + polyfil step.

It's currently importing from the semver NPM module, I'm using a light wrapper to select the 3 functions and namespace them before packaging them in the required IIFE to work with this odd build format.

First thing I did was try switching from `http-server` to `serve`.  Serve wont allow you to import from a parent directory (which is correct).  Had to switch it back because it would mean adjusting existing examples / directory structure.  Thought that would be best done separately.

epreston commented:
I would like to progress this.  Merging it into one file makes changing the build system and rewriting this very complex and risky.  If that is required, I can update this PR, but work off of a branch that is nicely separated.

willeastcott commented:
The `spine41.js` differs only in 3 places and the changes are not very significant. I genuinely don't think that there's a major risk in simply updating the original `spine.js`. As I've mentioned before, we want to do a wholesale refactor of `spine.js` and it would be problematic to have to do those updates to two remarkably similar files. A good code review plus a bit of testing should catch any issues.

epreston commented:
On the way.

epreston committed:
Combine files into one.

API differs before we can know which which file version this is supposed to support. The data file opened will determine which code paths are used.  This file may or may not be concatenated with the spine library that supports it.

https://github.com/playcanvas/playcanvas-spine/pull/73

epreston committed:
Update Gruntfile.js

update build system

epreston committed:
Update Prebuilt Files

epreston commented:
As requested:  rolled into one file.

The API differs before we can know which which file version this is supposed to support. The data file opened will determine which code paths are used.  This file may or may not be concatenated with the spine library that supports it.

It is using the length of the function signatures to guess which library has been concatenated to be able to do enough to be able to interrogate the skeleton file for a version.

willeastcott commented on Gruntfile.js:
```diff
@@ -1,16 +1,22 @@
     const src36 = [
-        'contrib/spine-ts/build/3.6/spine-core.js'
+        'contrib/spine-ts/build/3.6/spine-core.js',
+        'src/spine.js'
```
You can move this back up to `srcCommon`?

epreston replied:
Completed in: 4aab6ec

willeastcott commented on README.md:
```diff
@@ -12,13 +12,21 @@ Examples such as the Hero above can be found in the `examples` folder. To run th
 ### Editor
 
-Add the files `lib/playcanvas-spine.3.6.min.js` and `lib/spine.js` to your project. Note that a 3.8 version of the library is available too.
+Add the plugin matching the Spine version used to export the animations, i.e `build/playcanvas-spine.X.X.min.js` and the Playcanvas script `build/spine.js` to your project. The following plugins are available:
```
```suggestion
Add the plugin matching the Spine version used to export the animations, i.e `build/playcanvas-spine.X.X.min.js` and the PlayCanvas script `build/spine.js` to your project. The following plugins are available:
```

willeastcott commented on README.md:
```diff
@@ -12,13 +12,21 @@ Examples such as the Hero above can be found in the `examples` folder. To run th
 ### Engine-only
 
-Load the library script `lib/playcanvas-spine.3.6.min.js`. Then, add spine components to your entities as follows:
+Load the required library script, i.e `build/playcanvas-spine.X.X.min.js`. Then, add spine components to your entities as follows:
```
```suggestion
Load the required library script, i.e. `build/playcanvas-spine.X.X.min.js`. Then, add spine components to your entities as follows:
```

willeastcott commented on src/spine.js:
```diff
@@ -97,10 +97,13 @@ pc.extend(pc, function () {
 
         getImage: function () {
             return this._image;
+        },
```
```suggestion
        },

```

willeastcott commented on src/spine.js:
```diff
@@ -540,7 +569,9 @@ pc.extend(pc, function () {
     Spine.prototype.SubmitBatch = function (indexBase, indexCount, materialKey) {
         if (indexCount > 0) {
-            var mesh = new pc.Mesh();
+            // assumes Engine version 1.27+
```
I don't think you need this comment. Even if an engine before 1.27.0 was used (which is pretty unlikely!), it'd still be valid code because the parameter would simply be ignored.

epreston replied:
Completed in: 694e416

willeastcott commented on src/spine.js:
```diff
@@ -127,15 +130,37 @@ pc.extend(pc, function () {
         this.skeletonVersion = _skeletonData.version;
-        this._spine_3_6_0 = versionCompare(this.skeletonVersion, "3.6.0") <= 0;
-        this._spine_3_7_99 = versionCompare(this.skeletonVersion, "3.7.99") <= 0;
+        this._spine_3_6_0 = versionCompare(this.skeletonVersion, "3.6.0") <= 0; // version 3.6.0 or below
+        this._spine_3_7_99 = versionCompare(this.skeletonVersion, "3.7.99") <= 0; // version 3.7.99 or below
+        this._spine_4_1_X = semver.satisfies(semver.valid(semver.coerce(this.skeletonVersion)), '~4.1.23'); // version 4.1 family
```
I suspect you're apprehensive about breaking something, but can't we switch all version testing to `semver` and delete `versionCompare`? Seems strange to have two completely different mechanisms to check versions here.

epreston replied:
I am.  Ok, best approximation completed in 13dc31e

epreston committed:
Update Gruntfile.js

epreston committed:
Update README.md

Co-authored-by: Will Eastcott <will@playcanvas.com>

epreston committed:
Update README.md

Co-authored-by: Will Eastcott <will@playcanvas.com>

epreston committed:
Update src/spine.js

Co-authored-by: Will Eastcott <will@playcanvas.com>

epreston committed:
Update spine.js

epreston committed:
Update spine.js

Use semver for semver

epreston committed:
Update Builds

Update builds

epreston committed:
Update Gruntfile.js

update license banner

epreston committed:
Update package.json

increment package version

epreston committed:
update prebuilt modules

willeastcott commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "playcanvas-spine",
-  "version": "1.0.2",
+  "version": "1.1.0",
```
Let's bump version in a separate PR. I normally use the `npm version` command to bump package version instead of doing it by hand.

epreston replied:
Oh yes, for the tag and release commit. 

Does it do the lock file too or do you need to do `npm install` ?

One moment.

epreston replied:
Completed in f171916

epreston committed:
Update package.json

rewind package bump

epreston committed:
Update Gruntfile.js

ensure banner is dynamic

willeastcott commented on src/spine.js:
```diff
@@ -127,15 +85,37 @@ pc.extend(pc, function () {
+
+        // Compatibility queries
+        this.skeletonVersion = semver.valid(semver.coerce(_skeletonData.version));
+        this._spine_3_6_0 = semver.satisfies(this.skeletonVersion, '<3.6.0'); // version 3.6.0 or below
```
![image](https://github.com/playcanvas/playcanvas-spine/assets/697563/b513cb25-b7d9-4d8b-8c6d-655d4f2dc838)

epreston replied:
My comment is wrong.  It should be "below version 3.6.0"

The value used is tested on the behaviour of the function that was used before.

willeastcott replied:
But doesn't `versionCompare(this.skeletonVersion, "3.6.0") <= 0` say "less than or equal to 3.6.0"?

epreston replied:
You're right.  Ran a test with:

```
  console.log(`'3.5.0' ${versionCompare("3.5.0", "3.6.0") <= 0}`);
  console.log(`'3.6.0' ${versionCompare("3.6.0", "3.6.0") <= 0}`);
  console.log(`'3.6.1' ${versionCompare("3.6.1", "3.6.0") <= 0}`);
```

Updating now.

epreston replied:
completed in 29b6250

willeastcott commented on src/spine.js:
```diff
@@ -127,15 +85,37 @@ pc.extend(pc, function () {
+        // Compatibility queries
+        this.skeletonVersion = semver.valid(semver.coerce(_skeletonData.version));
+        this._spine_3_6_0 = semver.satisfies(this.skeletonVersion, '<3.6.0'); // version 3.6.0 or below
+        this._spine_3_7_99 = semver.satisfies(this.skeletonVersion, '<3.8.0'); // version 3.7.99 or below
```
And should this be:
`<=3.7.99`

epreston replied:
need to handle the case of `3.7.100` etc.  I'll update the comment.

epreston replied:
Here's the thing on this one.  I can't find where Spline editor ever produced files with versions this high, so I took this to mean "upper limit before 3.8".  Let's go with your suggestion.

epreston replied:
completed in 29b6250

epreston committed:
Update spine.js

adjust semver values

epreston committed:
Update prebuilt modules

willeastcott reviewed (approved):
Looks good to me. A bit of a marathon review, @epreston, but thanks for being patient with me! Really happy with this! 🚀

epreston commented:
I'm grateful for the time you have given to this.  Super keen to start the build system refactor.

mvaligursky commented:
I had a look and all this looks great, builds for me and example run.

I just wonder about the reason the generated files are larger then before.

Before:
![before](https://github.com/playcanvas/playcanvas-spine/assets/59932779/8e281802-46e6-4835-9ed3-a24736f1794e)

Now:
![now](https://github.com/playcanvas/playcanvas-spine/assets/59932779/1a9a887a-14c8-425e-a17a-ff75420be406)

willeastcott commented:
@mvaligursky I suspect that's due to the inclusion of the semver source. Feels like there should be an easier and cheaper way to do semver version comparisons, but maybe we can improve that in a follow up PR.

epreston commented:
> I had a look and all this looks great, builds for me and example run.
> 
> I just wonder about the reason the generated files are larger then before.
> 

yep.. consequence of combining things. 22 kb bloat.  The new build system cuts these files down by half.

> cheaper way to do semver version comparisons,

The Spline project does not apply file versioning in a way that makes sense outside of their own needs. 3.6 and 3.7 are as far away as a major revision.  With 4.1, they are not using a compliant version tag (at least in the validation files they provide) which forces 3rd parties to sanitise version comparisons. It's smart on their behalf.

mvaligursky commented:
Hi @epreston - would you have any time at all to perhaps add a support for spine 4.2 now that it's out?
